File: openacc/nestedparallel/lab2/nestedparallel.f90
module foo
contains

subroutine matvec(A,V,X,i,N)
        !$acc routine 
        implicit none
        integer, intent(in) :: i
        integer, intent(in) :: N
        real(4) :: A(N*N), V(N), X(N)
        real :: sum
        integer :: j
        sum = 0
        !$acc loop 
        do j=1,n
          sum = sum + a((i-1)*N+j)*v(j)
        enddo
        !$acc end loop

        X(i) = sum
end subroutine matvec
end module foo

program nestedparallel
        use openacc 
        use foo
        implicit none
        integer :: N = 20000
        integer :: i,j
        real(4), dimension(:), allocatable :: A
        real(4), dimension(:), allocatable :: V
        real(4), dimension(:), allocatable :: X
        real T1,T2,time
        real MFLOPS
        !#ifdef _OPENACC
          call acc_init(acc_device_nvidia)
        !#endif
        allocate(A(N*N))
        allocate(V(N))
        allocate(X(N))

        write(*,"(A,I5,A,I5,A,I5,A)") "Matrix Vector Multiplication of [",N,",",N,"] X [",N,"]"

        do i=1,N
          do j=1,N
            A((i-1)*N+j) = i+j
          enddo
          V(i) = i+j
        enddo
        
        CALL CPU_TIME(T1)
        !$acc data copy(A(0:N*N),V(0:N),X(0:N))
        !$acc parallel loop present(A(0:N*N),V(0:N),X(0:N)) 
        do i=1,N
          CALL matvec(A,V,X,i,N)
        enddo
        !$acc end parallel
        !$acc end data
        CALL CPU_TIME(T2)
        time = T2 - T1
        MFLOPS = (N*N) / (1E6 * time)
        print '("Execution Time: ",f6.3, " seconds ")', time
        print *, "MFLOPS/sec: ", MFLOPS
        
end program nestedparallel

File: openacc/nestedparallel/lab1/solution/nestedparallel.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <openacc.h>

#pragma acc routine 
void matvec(float * restrict v, float * restrict x, float * restrict a, int i, int n)
{
  float sum = 0;
  #pragma acc loop
  for (int j = 0; j < n; j++)
    sum += a[i*n+j]*v[j];

  x[i] = sum;
}
int main()
{
#ifdef _OPENACC
  acc_init(acc_device_nvidia);
#endif

  int N = 20000;
  int i,j;
  float *A, *V, *X;
  double start,end,time;
  double MFLOPS;
  A = (float*)malloc(sizeof(float)*N*N);
  V = (float*)malloc(sizeof(float)*N);
  X = (float*)malloc(sizeof(float)*N);

  printf("Matrix Vector Multiplication of [%d,%d] X [%d]\n ", N,N,N);

  for (i = 0; i < N; i++)
  {
    for (j = 0; j < N; j++)
      A[i*N+j] = i+j;
  
    V[i] = i+j;
  }
  start = clock();
  #pragma acc data copy(A[0:N*N],V[0:N],X[0:N])
  #pragma acc parallel loop present(A[0:N*N],V[0:N],X[0:N])
  for (i = 0; i < N; i++)
  {
    matvec(V,X,A,i,N);
  }
  end = clock();
  time = (end-start)/CLOCKS_PER_SEC;
  MFLOPS = (N*N) / (1E6 * time);
  printf("Execution: %f sec \n", time);
  printf("MFLOPS/sec : %f\n", MFLOPS);
  return 0;
}
